File: clients/algoliasearch-client-java/algoliasearch/src/main/java/com/algolia/transport/interceptors/LogInterceptor.java
package com.algolia.transport.interceptors;

import com.algolia.config.LogLevel;
import com.algolia.config.Logger;
import okhttp3.Interceptor;
import okhttp3.Response;
import okhttp3.logging.HttpLoggingInterceptor;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;

public class LogInterceptor implements Interceptor {
    private final HttpLoggingInterceptor logger;
    public LogInterceptor(Logger logger, LogLevel logLevel) {
        HttpLoggingInterceptor.Logger logr = logger != null
                ? toLogger(logger)
                : HttpLoggingInterceptor.Logger.DEFAULT;
        HttpLoggingInterceptor.Level level = toLevel(logLevel);
        this.logger = new HttpLoggingInterceptor(logr).setLevel(level);
    }

    public HttpLoggingInterceptor.Logger toLogger(@NotNull Logger logger) {
        return logger::log;
    }

    public HttpLoggingInterceptor.Level toLevel(@NotNull LogLevel logLevel) {
        switch (logLevel) {
            case NONE:
                return HttpLoggingInterceptor.Level.NONE;
            case BODY:
                return HttpLoggingInterceptor.Level.BODY;
            case BASIC:
                return HttpLoggingInterceptor.Level.BASIC;
            case HEADERS:
                return HttpLoggingInterceptor.Level.HEADERS;
            default:
                throw new UnsupportedOperationException("Unsupported LogLevel " + logLevel);
        }
    }

    @NotNull
    @Override
    public Response intercept(@NotNull Chain chain) throws IOException {
        return logger.intercept(chain);
    }
}

File: clients/algoliasearch-client-scala/src/main/scala/algoliasearch/internal/interceptor/LogInterceptor.scala
package algoliasearch.internal.interceptor

import algoliasearch.config.{LogLevel, Logger, Logging}
import okhttp3.logging.HttpLoggingInterceptor
import okhttp3.{Interceptor, Response}

/** Interceptor that logs HTTP requests and responses.
  *
  * @param logger
  *   Optional logger.
  * @param logLevel
  *   Log level.
  */
private[internal] class LogInterceptor(
    logging: Logging
) extends Interceptor {

  private val httpLogger: HttpLoggingInterceptor = {
    val logr: HttpLoggingInterceptor.Logger = logging.logger match {
      case Logger.Default => HttpLoggingInterceptor.Logger.DEFAULT
      case logger         => (message: String) => logger.log(message)
    }
    val level: HttpLoggingInterceptor.Level = logging.level match {
      case LogLevel.None    => HttpLoggingInterceptor.Level.NONE
      case LogLevel.Body    => HttpLoggingInterceptor.Level.BODY
      case LogLevel.Basic   => HttpLoggingInterceptor.Level.BASIC
      case LogLevel.Headers => HttpLoggingInterceptor.Level.HEADERS
      case _ =>
        throw new UnsupportedOperationException(
          s"Unsupported LogLevel ${logging.level}"
        )
    }
    new HttpLoggingInterceptor(logr).setLevel(level)
  }

  override def intercept(chain: Interceptor.Chain): Response =
    httpLogger.intercept(chain)
}
